packages/transaction-controller/src/helpers/ResimulateHelper.ts

```diff
@@ -1,31 +1,141 @@
+import { getPercentageChange } from '../utils/utils';
 
-const log = createModuleLogger(projectLogger, 'resimulate');
+const log = createModuleLogger(projectLogger, 'resimulate-helper');
 
 export const RESIMULATE_PARAMS = ['to', 'value', 'data'] as const;
 export const BLOCKAID_RESULT_TYPE_MALICIOUS = 'Malicious';
 export const VALUE_COMPARISON_PERCENT_THRESHOLD = 5;
 export const BLOCK_TIME_ADDITIONAL_SECONDS = 60;
+export const RESIMULATE_INTERVAL_MS = 3000;
 
 export type ResimulateResponse = {
   blockTime?: number;
   resimulate: boolean;
 };
 
+export type ResimulateHelperOptions = {
+  getTransactions: () => TransactionMeta[];
+  onTransactionsUpdate: (listener: () => void) => void;
+  simulateTransaction: (transactionMeta: TransactionMeta) => Promise<void>;
+};
+
+export class ResimulateHelper {
+  // Map of transactionId <=> timeoutId
+  readonly #timeoutIds: Map<string, NodeJS.Timeout> = new Map();
+
+  readonly #getTransactions: () => TransactionMeta[];
+
+  readonly #simulateTransaction: (
+    transactionMeta: TransactionMeta,
+  ) => Promise<void>;
+
+  constructor({
+    getTransactions,
+    simulateTransaction,
+    onTransactionsUpdate,
+  }: ResimulateHelperOptions) {
+    this.#getTransactions = getTransactions;
+    this.#simulateTransaction = simulateTransaction;
+
+    onTransactionsUpdate(this.#onTransactionsUpdate.bind(this));
+  }
+
+  #onTransactionsUpdate() {
+    const unapprovedTransactions = this.#getTransactions().filter(
+      (tx) => tx.status === TransactionStatus.unapproved,
+    );
+
+    const unapprovedTransactionIds = new Set(
+      unapprovedTransactions.map((tx) => tx.id),
+    );
+
+    // Combine unapproved transaction IDs and currently active resimulations
+    const allTransactionIds = new Set([
+      ...unapprovedTransactionIds,
+      ...this.#timeoutIds.keys(),
+    ]);
+
+    allTransactionIds.forEach((transactionId) => {
+      const transactionMeta = unapprovedTransactions.find(
+        (tx) => tx.id === transactionId,
+      );
+
+      if (transactionMeta && transactionMeta.isActive) {
+        this.#start(transactionMeta);
+      } else {
+        this.#stop({
+          id: transactionId,
+          isActive: false,
+        } as unknown as TransactionMeta);
+      }
+    });
+  }
+
+  #start(transactionMeta: TransactionMeta) {
+    const { id: transactionId } = transactionMeta;
+    if (!transactionMeta.isActive || this.#timeoutIds.has(transactionId)) {
+      return;
+    }
+
+    const listener = () => {
+      // eslint-disable-next-line promise/catch-or-return
+      this.#simulateTransaction(transactionMeta)
+        .catch((error) => {
+          /* istanbul ignore next */
+          log('Error during transaction resimulation', error);
+        })
+        .finally(() => {
+          // Schedule the next execution
+          if (this.#timeoutIds.has(transactionId)) {
+            const timeoutId = setTimeout(listener, RESIMULATE_INTERVAL_MS);
```

Comment:
It's very rare we'll have more than one transaction, but worth confirming this means the updates for different transactions won't be in sync.

This should be fine as the simulation isn't a batch request.